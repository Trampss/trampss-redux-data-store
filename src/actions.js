const scope = action => `@trampss/${action.toUpperCase()}`

export const SET = prefix => scope(`SET_${prefix}`)
export const set = prefix => payload => ({ type: SET(prefix), payload })

export const RESET = prefix => scope(`RESET_${prefix}`)
export const reset = prefix => () => ({ type: RESET(prefix) })

export const ADD = prefix => scope(`ADD_${prefix}`)
export const add = prefix => payload => ({ type: ADD(prefix), payload })

export const UPDATE = prefix => scope(`UPDATE_${prefix}`)
export const update = prefix => payload => ({ type: UPDATE(prefix), payload })

export const REMOVE = prefix => scope(`REMOVE_${prefix}`)
export const remove = prefix => key => ({ type: REMOVE(prefix), payload: key })

export const ADD_OR_UPDATE = prefix => scope(`ADD_OR_UPDATE_${prefix}`)
export const addOrUpdate = prefix => payload => ({ type: ADD_OR_UPDATE(prefix), payload })
