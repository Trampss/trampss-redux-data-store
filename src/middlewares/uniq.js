import { SET, RESET } from '../actions'

export const initState = { data: undefined, initialized: false }

const reducer = (/* key */) => prefix =>
  (state = initState, { type, payload } = {}) => {
    switch (type) {
      case SET(prefix):
        return {
          data: payload,
          initialized: true,
        }
      case RESET(prefix):
        return initState
      default:
        return state
    }
  }

export default key => prefix => (ctx = {}) => ({
  ...ctx,
  state: reducer(key)(prefix)(ctx.state, ctx.action),
})
