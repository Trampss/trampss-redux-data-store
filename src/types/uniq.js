import uniq from '../middlewares/uniq'

export default {
  middlewares: [uniq],
  actions: ['set', 'reset', 'update'],
  selectors: ['getState', 'isInitialized', 'get'],
}
