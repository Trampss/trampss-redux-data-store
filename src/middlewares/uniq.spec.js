/* eslint-env jest */
import { set, reset, update } from '../actions'
import uniq from './uniq'

const prefix = 'testPrefix'
const Element = code => ({ code, some: 'other', infos: code })
const state = {
  data: Element('elm'),
  initialized: true,
}

describe('middlewares/uniq', () => {
  // Use the factory to create a new reducer named 'testPrefix'
  // It uses 'code' as key in givent elements
  const testPrefix = uniq('code')(prefix)

  it('should initialize', () => expect(testPrefix()).toMatchSnapshot())

  it('should set element [elm2]', () => expect(
    testPrefix({
      state,
      action: set(prefix)(Element('elm2')),
    }),
  ).toMatchSnapshot())

  it('should reset state', () => expect(
    testPrefix({
      state,
      action: reset(prefix)(),
    }),
  ).toMatchSnapshot())

  it('should update element', () => expect(
    testPrefix({
      state,
      action: update(prefix)({ some: 'other 2', modifi: 'cation' }),
    }),
  ).toMatchSnapshot())
})
