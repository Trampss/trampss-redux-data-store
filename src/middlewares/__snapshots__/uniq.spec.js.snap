// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`middlewares/uniq should initialize 1`] = `
Object {
  "state": Object {
    "data": undefined,
    "initialized": false,
  },
}
`;

exports[`middlewares/uniq should reset state 1`] = `
Object {
  "action": Object {
    "type": "@trampss/RESET_TESTPREFIX",
  },
  "state": Object {
    "data": undefined,
    "initialized": false,
  },
}
`;

exports[`middlewares/uniq should set element [elm2] 1`] = `
Object {
  "action": Object {
    "payload": Object {
      "code": "elm2",
      "infos": "elm2",
      "some": "other",
    },
    "type": "@trampss/SET_TESTPREFIX",
  },
  "state": Object {
    "data": Object {
      "code": "elm2",
      "infos": "elm2",
      "some": "other",
    },
    "initialized": true,
  },
}
`;

exports[`middlewares/uniq should update element 1`] = `
Object {
  "action": Object {
    "payload": Object {
      "modifi": "cation",
      "some": "other 2",
    },
    "type": "@trampss/UPDATE_TESTPREFIX",
  },
  "state": Object {
    "data": Object {
      "code": "elm",
      "infos": "elm",
      "modifi": "cation",
      "some": "other 2",
    },
    "initialized": true,
  },
}
`;
